samza-test/src/main/java/org/apache/samza/test/framework/stream/CollectionStream.java

```diff
@@ -0,0 +1,180 @@
+
+  private String streamName;
+  private CollectionStreamSystem system;
+  private Map<Integer, Iterable<T>> partitions;
+  private Map<String, String> streamConfig;
+  private static final String STREAM_TO_SYSTEM = "streams.%s.samza.system";
+  private static final String DEFAULT_SYSTEM = "_default_test_system_";
+
+  /**
+   * Constructs a new CollectionStream with multiple empty partitions from specified components.
+   * @param streamName represents name of the stream
+   * @param partitionCount represents number of partitions, each of these partitions will be empty
+   */
+  private CollectionStream(String streamName, Integer partitionCount) {
+    this.streamName = streamName;
+    this.streamConfig = new HashMap<>();
+    this.partitions = new HashMap<>();
+    this.system = CollectionStreamSystem.create(DEFAULT_SYSTEM);
+    for (int i = 0; i < partitionCount; i++) {
+      partitions.put(i, new ArrayList<>());
+    }
+  }
+
+  /**
+   * Constructs a new CollectionStream with single partition from specified components.
+   * @param streamName represents name of the stream
+   * @param partition represents the messages that the stream will be intialized with, default partitionId for the
+   *                  this single partition stream is 0
+   */
+  private CollectionStream(String streamName, Iterable<T> partition) {
+    this.streamName = streamName;
+    this.streamConfig = new HashMap<>();
+    partitions = new HashMap<>();
+    this.system = CollectionStreamSystem.create(DEFAULT_SYSTEM);
+    partitions.put(0, partition);
+  }
+
+  /**
+   * Constructs a new CollectionStream with multiple partitions from specified components.
+   * @param streamName represents name of the stream
+   * @param partitions represents the partition state, key of the map represents partitionId and value represents
+   *                   the messages that partition will be initialized with
+   */
+  private CollectionStream(String streamName, Map<Integer, ? extends Iterable<T>> partitions) {
+    this.streamName = streamName;
+    this.streamConfig = new HashMap<>();
+    this.partitions = new HashMap<>();
+    partitions.entrySet().forEach(entry -> {
+        this.partitions.put(entry.getKey(), entry.getValue());
+      });
+  }
+
+  public CollectionStreamSystem getCollectionStreamSystem() {
+    return system;
+  }
+
+  public Map<Integer, Iterable<T>> getInitPartitions() {
+    return partitions;
+  }
+
+  public String getStreamName() {
+    return streamName;
+  }
+
+  public Map<String, String> getStreamConfig() {
+    return streamConfig;
+  }
+
+  /**
+   * Associates the stream with a {@link CollectionStreamSystem} and generates the configuration for it.
+   *
+   * @param system represents the {@link CollectionStreamSystem} that stream is associated with
+   * @return the {@link CollectionStream} associated with {@link CollectionStreamSystem}
+   */
+  public CollectionStream<T> from(CollectionStreamSystem system) {
```

Comment:
I think we should have a way to hide this. The user can just specify the system name and we will create the CollectionStreamSystem behind the scene. It's kind of weird for low-level users to care about this.

Reply:
sure